BossMod/QuestBattle/QuestBattleDirector.cs

```diff
@@ -0,0 +1,375 @@
+        }
+    }
+
+    public void Update(AIHints hints)
+    {
+        if (!Enabled || Paused || bmm?.ActiveModule?.StateMachine.ActivePhase != null)
+            return;
+
+        var player = World.Party.Player();
+        if (player == null)
+            return;
+
+        if (HaveTarget(player, hints))
+        {
+            if (!_combatFlag)
+                OnPlayerEnterCombat(player);
+
+            _combatFlag = true;
+        }
+        else if (_combatFlag)
+        {
+            OnPlayerExitCombat(player);
+            _combatFlag = false;
+        }
+
+        if (CurrentModule != null)
+        {
+            CurrentModule.Update();
+            var obj = CurrentModule.CurrentObjective;
+            if (obj == null)
+            {
+                if (CurrentObjective != null)
+                {
+                    ObjectiveCleared.Fire(CurrentObjective);
+                    CurrentObjective = null;
+                }
+            }
+            else if (obj != CurrentObjective)
+            {
+                ObjectiveChanged.Fire(obj);
+                CurrentObjective = obj;
+            }
+        }
+
+        if (CurrentObjective != null)
+            MoveNext(player, CurrentObjective, hints);
+    }
+
+    private void MoveNext(Actor player, QuestObjective objective, AIHints hints)
+    {
+        if (CurrentWaypoints.Count == 0 || AIController.InCutscene)
+            return;
+
+        if (_config.ShowWaypoints)
+            DrawWaypoints(player.PosRot.XYZ());
+
+        var nextwp = CurrentWaypoints[0];
+        var playerPos = player.PosRot.XYZ();
+        var direction = nextwp.Position - playerPos;
+        if (direction.XZ().Length() < Tolerance)
+        {
+            if (nextwp.SpecifiedInPath)
+                CurrentObjectiveNavigationProgress++;
+
+            CurrentWaypoints.RemoveAt(0);
+            if (CurrentWaypoints.Count == 0)
+                CurrentModule?.OnNavigationComplete();
+            MoveNext(player, objective, hints);
+            return;
+        }
+        else
+        {
+            Dash(player, direction, hints);
+            hints.ForcedMovement = direction;
+        }
+    }
+
+    private void DrawWaypoints(Vector3 playerPos)
+    {
+        var start = playerPos;
+        var current = true;
+        foreach (var wp in CurrentWaypoints)
+        {
+            Camera.Instance?.DrawWorldLine(start, wp.Position, current ? ArenaColor.Safe : ArenaColor.Danger);
+            current = false;
+            start = wp.Position;
+        }
+    }
+
+    public static bool HaveTarget(Actor player, AIHints hints) => player.InCombat || hints.PriorityTargets.Any(x => hints.Bounds.Contains(x.Actor.Position - hints.Center));
+
+    enum ActionsProhibitedStatus : uint
+    {
+        OutOfTheAction = 2109,
+        InEvent = 1268,
+    }
+
+    private void Dash(Actor player, Vector3 direction, AIHints hints)
```

Comment:
This sounds like some autorotation / action tweaks / ai level rather than quest modules?

Reply:
i'm not discounting the possibility, i just don't know where it belongs in the codebase :)